### PR TITLE
Implement API for Shulkers, ShulkerBullets and ShulkerBoxes

### PR DESCRIPTION
**Pull request by RedNesto**
[SpongeAPI](https://github.com/SpongePowered/SpongeAPI/pull/1903) | **SpongeCommon**

This PR implements everything added to its SpongeAPI counterpart.

I hope I did all the DataAPI stuff right.

Reviews are welcome.

_Only tested with SpongeVanilla for now._

[Test Plugin](https://gist.github.com/RedNesto/6fdd0d1e8ce66a52374432e1544c43ba)

### CODE REVIEW

File: src/main/java/org/spongepowered/common/data/processor/data/entity/EntityTargetedEntityDataProcessor.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+public final class EntityTargetedEntityDataProcessor extends AbstractEntitySingleDataProcessor<Entity, Optional<org.spongepowered.api.entity.Entity>,
+        OptionalValue<org.spongepowered.api.entity.Entity>, TargetedEntityData, ImmutableTargetedEntityData> {
+
+    public EntityTargetedEntityDataProcessor() {
+        super(Entity.class, Keys.TARGETED_ENTITY);
+    }
+
+    @Override
+    protected boolean set(Entity dataHolder, Optional<org.spongepowered.api.entity.Entity> value) {
+        if (!(dataHolder instanceof IEntityTargetingEntity))
```
Comment (XakepSDK) [inline src/main/java/org/spongepowered/common/data/processor/data/entity/EntityTargetedEntityDataProcessor.java]: Braces

File: src/main/java/org/spongepowered/common/data/processor/data/entity/EntityTargetedEntityDataProcessor.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+            return false;
+
+        ((IEntityTargetingEntity) dataHolder).setTargetedEntity((Entity) value.orElse(null));
+
+        return true;
+    }
+
+    @Override
+    protected Optional<Optional<org.spongepowered.api.entity.Entity>> getVal(Entity dataHolder) {
+        if (!(dataHolder instanceof IEntityTargetingEntity))
```
Comment (XakepSDK) [inline src/main/java/org/spongepowered/common/data/processor/data/entity/EntityTargetedEntityDataProcessor.java]: Braces

File: src/main/java/org/spongepowered/common/data/processor/data/entity/ShulkerDyeColorDataProcessor.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    @Override
+    protected Value<DyeColor> constructValue(DyeColor actualValue) {
+        return SpongeValueFactory.getInstance().createValue(Keys.DYE_COLOR, actualValue, DyeColors.PURPLE);
+    }
+
+    @Override
+    protected boolean set(EntityShulker container, DyeColor value) {
+        if (!(container instanceof IMixinShulker))
```
Comment (XakepSDK) [inline src/main/java/org/spongepowered/common/data/processor/data/entity/ShulkerDyeColorDataProcessor.java]: Braces

File: src/main/java/org/spongepowered/common/data/processor/data/entity/ShulkerDyeColorDataProcessor.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+            return false;
+
+        ((IMixinShulker) container).setColor(value);
+        return true;
+    }
+
+    @Override
+    protected Optional<DyeColor> getVal(EntityShulker container) {
+        if (!(container instanceof IMixinShulker))
```
Comment (XakepSDK) [inline src/main/java/org/spongepowered/common/data/processor/data/entity/ShulkerDyeColorDataProcessor.java]: Braces

File: src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntityShulkerBullet.java [hunk 5]
```diff
@@ -24,11 +24,115 @@
...
+            this.targetUniqueId = entity.getUniqueID();
+        } else {
+            this.targetUniqueId = null;
+        }
+    }
+
+    @Override
+    public void supplyVanillaManipulators(List<DataManipulator<?, ?>> manipulators) {
+        super.supplyVanillaManipulators(manipulators);
+        manipulators.add(new SpongeTargetedEntityData());
```
Comment (ImMorpheus) [inline src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntityShulkerBullet.java]: This is wrong. It's always returning an empty target.

**Review by ImMorpheus — DISMISSED** [review]

File: src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntityShulkerBullet.java [hunk 6]
```diff
@@ -24,11 +24,115 @@
...
+    }
+
+    @Override
+    public OptionalValue<Entity> target() {
+        return new SpongeOptionalValue<>(Keys.TARGETED_ENTITY, Optional.ofNullable((Entity) this.target));
+    }
+
+    @Override
+    public ProjectileSource getShooter() {
+        if (projectileSource != null) {
```
Comment (ImMorpheus) [inline src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntityShulkerBullet.java]: this

**Review by ImMorpheus — CHANGES_REQUESTED** [review]

**Review by ImMorpheus — CHANGES_REQUESTED** [review]
As in the signsource PR, add the test plugin source as a proper test plugin.

**Review by ImMorpheus — APPROVED** [review]

**Pull request merged**